File: Week_05/Practicum/Next Greater Element II.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        std::stack<std::pair<int, int>> stack;
        std::vector<int> ans(nums.size(), -1);
        for (int i = 0; i < nums.size(); ++i) {
            while (!stack.empty() && stack.top().first < nums[i]) {
                ans[stack.top().second] = nums[i];
                stack.pop();
            }

            stack.emplace(nums[i], i);
        }

        for (int i = 0; i < nums.size(); ++i) {
            while (!stack.empty() && stack.top().first < nums[i]) {
                ans[stack.top().second] = nums[i];
                stack.pop();
            }
        }
        
        return ans;
    }
};

class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        vector<int> res(n, -1);
        stack<int> st;

        for(int i = 0; i < 2 * n; i++) {
            while(!st.empty() && nums[st.top()] < nums[i % n]) {
                res[st.top()] = nums[i % n];
                st.pop();
            }
            st.push(i % n);
        }
        return res;
    }
};

File: Week_05/Practicum/Next Greater Element II.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        # 1 5 4 3 2 5 1 6
        # Keep monotically decrease stack, that way when we reach bigger number than
        # the top, we can pop all lesser elements

        ans = [-1 for _ in nums]

        stack = []

        # We will do 2 passes, since it is circular
        for _ in range(2):
            for i, num in enumerate(nums):

                while stack and stack[-1][1] < num:
                    ans[stack[-1][0]] = num
                    stack.pop()

                stack.append((i, num))

        return ans
            